Vetrov/statistics epic 2 3

Konstantin-vetro opened the pull request:
Выбранная архитектура: MVVM
Способ вёрстки: с помощью кода
ссылка на доску: https://github.com/users/Skoch33/projects/2

Konstantin-vetro committed:
удалил предыдущую реализацию, чтобы изменения были как на доске

Konstantin-vetro committed:
Сверстал ImageView фото профиля;
Сверстал NameLabel имени пользователя;
Сверстал DescriptionLabel, резюме пользователя;
Сверстал UserInfoButton, кнопки сайта пользователя;
Сверстал UserCollectionButton, кнопка просмотра коллекции NFT пользователя;

Konstantin-vetro committed:
отобразил данные пользователя в методе showUser()

Konstantin-vetro committed:
создал WebViewViewController с индикатором загрузки страницы

Konstantin-vetro committed:
реализовал переход на сайт пользователя с помощью WebView

Konstantin-vetro committed:
Реализовал переход на страницу коллекции NFT пользователя

Konstantin-vetro committed:
произвел рефакторинг на MVVM добавив viewModel

Konstantin-vetro committed:
на экране статистики передал данные о пользователе во вьюмодель пользователя

Konstantin-vetro committed:
поменял бандл-идентификатор в своем проекте

Konstantin-vetro committed:
добавил удаление кэша из памяти и сделал обновление аватарки пользователя асинхронно т.к. иногда не показывалась аватарка

taneev commented on FakeNFT/Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -4,11 +4,13 @@
-        let image = UIImageView(image: pic)
+        let placeholder = UIImage(systemName: "person.crop.circle.fill")
+        let image = UIImageView(image: placeholder)
         image.layer.cornerRadius = 35
```
Может быть имеет смысл тут и ниже (строки 120-121) использовать одну и ту же константу, чтобы при любом возможном будущем изменении размера форма картинки точно оставалсь круглой?

Konstantin-vetro replied:
placeholder добавил везде где требуется по умолчанию

taneev commented on FakeNFT/Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -19,15 +21,13 @@ final class UserCardViewController: UIViewController {
-        label.text = "Joaquin Phoenix"
         return label
     }()
     
```
тут в пустой строке остались пробелы (и не только тут, во многих местах). Можно такие места чистить автоматически, если в настройках Xcode включить настройку чистки от пробелов пустых строк ([ ] Including whitespace-only lines)

Konstantin-vetro replied:
включил в Xcode настройку чистки от пробелов пустых строк

taneev commented on FakeNFT/Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -47,31 +47,45 @@ final class UserCardViewController: UIViewController {
         super.viewDidLoad()
         setupNavigationBar()
         setupUI()
+        showUser()
```
Вот этот момент для меня самый неочевидный. С одной стороны, на этом экране нет никаких точек ввода данных пользователем, и никакой логики во viewModel'и, поэтому наверное такой вариант - это ок. Но в целом, как я понял идею, viewModel обычно говорит UI, что и в какой момент показывать. Т.е. в более сложном случае я бы тут просто сообщил viewModel'и, что view загрузилась (через вызов соответствующего метода), а уже viewModel через биндинги сообщила бы viewController'у, какие данные нужно отобразить в ответ на это событие. Но т.к. тут по сути нет никакой логики, то может быть такое усложнение было бы действительно излишним. Наверное это просто комментарий к размышлению)

Konstantin-vetro replied:
тоже хотел сделать это через bind но тут как таковая логика почти отсутствует, поэтому просто обновил значения через viewModel

taneev commented on FakeNFT/Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -47,31 +47,45 @@ final class UserCardViewController: UIViewController {
     @objc
     private func goToWebView() {
-        //TODO: вставить переход на webview
+        guard let userURL = viewModel.userWebsiteUrl else { return}
```
после return пробельчика не хватает
{ return }

Konstantin-vetro replied:
исправил

taneev commented on FakeNFT/Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -126,4 +141,22 @@ final class UserCardViewController: UIViewController {
+        cache.clearDiskCache()
+        
+        userImageView.kf.indicatorType = .activity
+        DispatchQueue.main.async {
```
Вот тут я не совсем понял идею с вызовом DispatchQueue.main.async. Ты вроде бы гарантированно уже находишься в главном потоке, и плюс вызов  kf.setImage сам по себе реализует асинхронный запрос. Точно ли здесь нужен вызов DispatchQueue.main.async?

Konstantin-vetro replied:
согласен, убрал асинхронный поток

taneev reviewed (commented):
Код мне понравился, приятно читать, все понятно. Оставил пару небольших вопросов и комментариев - может быть у тебя появятся какие-то идеи по этому поводу.
Может быть нужно еще обновить StatisticsFlowDecomposition.md - проставить там фактические трудозатраты?

Konstantin-vetro committed:
обновил StatisticsFlowDecomposition.md - проставил там фактические трудозатраты за 2 эпика

Konstantin-vetro committed:
обновил марки
placeholder добавил везде где требуется по умолчанию
поставил пробел в методе goToWebView после return
в методе setupUI массив subviews сделал более читаемым
убрал асинхронное выполнение аватарки пользователя в методе showUser

Konstantin-vetro committed:
Создал универсальную АлертМодель для переиспользования в коде

Konstantin-vetro committed:
провел рефакторинг кода основываясь но универсальном алерте, теперь когда отсутствует соединение с интернетом всплывает окно алерта

LizaIlinykh commented on FakeNFT/Statistics/Rating/AlertPresenter.swift:
```diff
@@ -0,0 +1,53 @@
+    private var title: String?
+    private var message: String?
+    private var actions: [AlertActionModel]
+    private var style: UIAlertController.Style
```
все свойства можно сделать константами

Konstantin-vetro replied:
исправил

LizaIlinykh commented on FakeNFT/Statistics/UserCard/UserCardViewModel.swift:
```diff
@@ -0,0 +1,50 @@
+    
+    var nftsID: [String] {
+        return user?.nfts ?? []
+    }
```
ты используешь вью модель как хранилище данных, откуда вью сама берет нужные ей данные. вью сама не должна обрабатывать события , она должна сообщить вью модели о событии, вью модель должна решить как это событие обработать и обновить вью через биндинг. сейчас у твоей вью модели нет ни одного метода для обработки событий, хотя события на этом экране есть, например : viewDidLoad, didTapInfoAboutUserButton, didTapNFTsCollectionButton. Вот как минимум эти три события должна обработать вью модель

Konstantin-vetro replied:
попробовал реализовать обработку событий через вьюмодель

Konstantin-vetro committed:
все свойства сделал константами

Konstantin-vetro committed:
cогласно замечаниям ревьювера реализовал обработку событий перехода на сайт и на экран коллекций пользователя через вьюмодель
